client-java/instrumentation/src/main/java/org/evomaster/client/java/instrumentation/coverage/methodreplacement/thirdpartyclasses/JacksonObjectMapperClassReplacement.java

```diff
@@ -35,65 +33,118 @@ protected String getNameOfThirdPartyTargetClass() {
 
     @Replacement(replacingStatic = false,
             type = ReplacementType.TRACKER,
-            id = "Jackson_ObjectMapper_readValue_InputStream_class",
+            id = "Jackson_ObjectMapper_readValue_InputStream_Generic_class",
             usageFilter = UsageFilter.ANY,
             category = ReplacementCategory.EXT_0)
     public static <T> T readValue(Object caller, InputStream src, Class<T> valueType) throws Throwable {
         Objects.requireNonNull(caller);
 
         ClassToSchema.registerSchemaIfNeeded(valueType);
 
-        /*
-            To be able to check the taint, we need to read the whole stream.
+        String content = new BufferedReader(
+                new InputStreamReader(src, Charset.defaultCharset()))
+                .lines()
+                .collect(Collectors.joining(System.lineSeparator()));
 
-            TODO: check if it has side-effects
-         */
+        JsonTaint.handlePossibleJsonTaint(content, valueType);
+
+        src = new ByteArrayInputStream(content.getBytes());
+
+        Method original = getOriginal(singleton, "Jackson_ObjectMapper_readValue_InputStream_Generic_class", caller);
+
+        try {
+            return (T) original.invoke(caller, src, valueType);
+        } catch (IllegalAccessException e) {
+            throw new RuntimeException(e);
+        } catch (InvocationTargetException e) {
+            throw e.getCause();
+        }
+    }
+
+    @Replacement(replacingStatic = false,
+            type = ReplacementType.TRACKER,
+            id = "Jackson_ObjectMapper_readValue_InputStream_TypeReference_class",
+            usageFilter = UsageFilter.ANY,
+            category = ReplacementCategory.EXT_0)
+    public static <T> T readValue(Object caller, InputStream src,
+                                  @ThirdPartyCast(actualType = "  com.fasterxml.jackson.databind.JavaType") Object valueType)
+            throws Throwable {
+        Objects.requireNonNull(caller);
+
+        Class<?> typeClass = (Class) valueType.getClass().getMethod("getRawClass").invoke(valueType);
 
         String content = new BufferedReader(
                 new InputStreamReader(src, Charset.defaultCharset()))
                 .lines()
                 .collect(Collectors.joining(System.lineSeparator()));
 
-        JsonTaint.handlePossibleJsonTaint(content,valueType);
+        ClassToSchema.registerSchemaIfNeeded(typeClass);
+        JsonTaint.handlePossibleJsonTaint(content, typeClass);
 
         src = new ByteArrayInputStream(content.getBytes());
```

Comment:
there is quite a bit of code here that is the same as previous function. should be possibly refactored in a private method to avoid code duplication

Reply:
but, to save time, i ll do this refactoring.

Reply:
Thank you.